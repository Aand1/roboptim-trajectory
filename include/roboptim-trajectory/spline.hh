// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Class Spline declaration.
 */

#ifndef ROBOPTIM_TRAJECTORY_SPLINE_HH
# define ROBOPTIM_TRAJECTORY_SPLINE_HH
# include <roboptim-trajectory/trajectory.hh>

# include <roboptim-trajectory/fwd.hh>

class bspline;

namespace roboptim
{
  class Spline : public Trajectory<4>
  {
  public:
    Spline (size_type, const vector_t&, int, int, int) throw ();
    virtual ~Spline () throw ();

    virtual vector_t operator () (double) const throw ();

    virtual vector_t derivative (double x, size_type order) const throw ();

    virtual jacobian_t variationConfigWrtParam (double t) const throw ();
    virtual jacobian_t variationDerivWrtParam (double t, size_type order)
      const throw ();
    virtual value_type singularPointAtRank (size_type rank) const;
    virtual vector_t derivBeforeSingularPoint (size_type rank, size_type order)
      const;
    virtual vector_t derivAfterSingularPoint (size_type rank, size_type order)
      const;

  private:
    bspline* spline_;
  };
} // end of namespace roboptim.

#endif //! ROBOPTIM_TRAJECTORY_TRAJECTORY_HH
