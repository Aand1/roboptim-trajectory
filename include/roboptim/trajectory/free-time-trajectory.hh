// Copyright (C) 2009 by Florent Lamiraux, Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_TRAJECTORY_FREETIMETRAJECTORY_HH
# define ROBOPTIM_TRAJECTORY_FREETIMETRAJECTORY_HH
# include <roboptim/trajectory/trajectory.hh>

namespace roboptim
{
  template <unsigned DerivabilityOrder>
  class FreeTimeTrajectory : public Trajectory<DerivabilityOrder>
  {
  public:
    /// Constructor with fixed definition interval trajectory
    ///
    /// \param traj trajectory defining this one by reparameterization
    /// \param s time scale
    FreeTimeTrajectory (const Trajectory<DerivabilityOrder>& traj, double s) throw ();
    virtual ~FreeTimeTrajectory () throw ();

    virtual vector_t operator () (double) const throw ();

    virtual vector_t derivative (double x, size_type order) const throw ();

    virtual jacobian_t variationConfigWrtParam (double t) const throw ();
    virtual jacobian_t variationDerivWrtParam (double t, size_type order)
      const throw ();
    virtual value_type singularPointAtRank (size_type rank) const;
    virtual vector_t derivBeforeSingularPoint (size_type rank, size_type order)
      const;
    virtual vector_t derivAfterSingularPoint (size_type rank, size_type order)
      const;

    /// Get index of time scaling parameter in vector
    double timeScale () const throw ();

    virtual std::ostream& print (std::ostream&) const throw ();
  private:
    double scaleTime (double) const throw ();

    /// Fixed time trajectory
    const Trajectory<DerivabilityOrder>& trajectory_;
    /// Store time scaling parameter \f$\textbf{p}_{m+1}\f$.
    double timeScale_;
  };
} // end of namespace roboptim.

# include <roboptim/trajectory/free-time-trajectory.hxx>
#endif //! ROBOPTIM_TRAJECTORY_FREETIMETRAJECTORY_HH
