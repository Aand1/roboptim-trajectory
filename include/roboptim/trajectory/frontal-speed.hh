// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_TRAJECTORY_FRONTAL_SPEED_HH
# define ROBOPTIM_TRAJECTORY_FRONTAL_SPEED_HH
# include <boost/shared_ptr.hpp>

# include <roboptim/core/derivable-function.hh>
# include <roboptim/trajectory/fwd.hh>

namespace roboptim
{
  template <typename T>
  class FrontalSpeed : public DerivableFunction
  {
  public:
    FrontalSpeed (const T& spline) throw ();
    ~FrontalSpeed () throw ();

  protected:
    void impl_compute (result_t& res, const argument_t& t) const throw ();
    void impl_gradient (gradient_t& grad, const argument_t& t, size_type i)
      const throw ();
  private:
    const T& trajectory_;
  };
} // end of namespace roboptim.

# include <roboptim/trajectory/frontal-speed.hxx>
#endif //! ROBOPTIM_TRAJECTORY_FRONTAL_SPEED_HH
