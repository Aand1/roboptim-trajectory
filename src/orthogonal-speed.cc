// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include <roboptim/trajectory/sys.hh>

#include <cmath>
#include <roboptim/trajectory/orthogonal-speed.hh>

namespace roboptim
{
namespace trajectory
{
  OrthogonalSpeed::OrthogonalSpeed ()
    : DerivableFunction (2 * 3, 1, "orthogonal speed")
  {}

  OrthogonalSpeed::~OrthogonalSpeed ()
  {}

  void
  OrthogonalSpeed::impl_compute (result_ref res, const_argument_ref x) const
  {
    const value_type& theta = x[2];

    const value_type& xdot = x[3 + 0];
    const value_type& ydot = x[3 + 1];
    res[0] = std::cos (theta) * ydot - std::sin (theta) * xdot;
  }


  void
  OrthogonalSpeed::impl_gradient (gradient_ref grad, const_argument_ref arg, size_type)
    const
  {
    //const value_type& x = arg[0];
    //const value_type& y = arg[1];
    const value_type& theta = arg[2];

    const value_type& dx = arg[3 + 0];
    const value_type& dy = arg[3 + 1];
    //const value_type& dtheta = arg[3 + 2];

    grad.setZero();
    // 0, 1, 5 components are null.
    grad[2] = -1. * (std::cos (theta) * dx + std::sin (theta) * dy);
    grad[3] = -std::sin (theta);
    grad[4] = std::cos (theta);
  }
} // end of namespace trajectory.
} // end of namespace roboptim.
