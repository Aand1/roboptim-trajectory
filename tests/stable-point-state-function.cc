// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include "common.hh"

#include <roboptim/core/io.hh>
#include <roboptim/core/finite-difference-gradient.hh>

#include <roboptim/trajectory/free-time-trajectory.hh>
#include <roboptim/trajectory/frontal-speed.hh>
#include <roboptim/trajectory/orthogonal-speed.hh>
#include <roboptim/trajectory/spline.hh>
#include <roboptim/trajectory/stable-point-state-function.hh>

using namespace roboptim;

int run_test ()
{
  const unsigned orderMax = 1;
  Spline::vector_t params (12);

  // Initial position.
  params[0] = 0.,  params[1] = 0., params[2] = 0.;
  // Control point 1.
  params[3] = 25.,  params[4] = 100., params[5] = 0.;
  // Control point 2.
  params[6] = 75.,  params[7] = 0., params[8] = 0.;
  // Final position.
  params[9] = 100., params[10] = 100., params[11] = 0.;

  Spline::vector_t fttParams = addScaleToParameters (params, 1.5);

  Spline::interval_t timeRange = Spline::makeInterval (0., 4.);
  Spline spline (timeRange, 3, params, "before");

  typedef FreeTimeTrajectory<Spline::derivabilityOrder> freeTimeTraj_t;
  freeTimeTraj_t ftt (spline, 2.);

  for (unsigned i = 0; i < 10; ++i)
    {
      const StableTimePoint timePoint = i / 10. * tMax;
      const double t = timePoint.getTime (ftt.timeRange ());

      boost::shared_ptr<DerivableFunction> frontalSpeed (new FrontalSpeed ());
      StablePointStateFunction<Spline> stateFunction
	(ftt, frontalSpeed, timePoint, orderMax);

      boost::shared_ptr<DerivableFunction> orthogonalSpeed
	(new OrthogonalSpeed ());
      StablePointStateFunction<Spline> orthoStateFunction
	(ftt, orthogonalSpeed, timePoint, orderMax);

      std::cout << "State cost evaluation:" << std::endl
		<< stateFunction (fttParams) << std::endl
		<< "State cost gradient:" << std::endl
		<< stateFunction.gradient (fttParams) << std::endl
		<< "Trajectory state (splitted):" << std::endl;
      for (unsigned o = 0; o <= orderMax; ++o)
	std::cout << ftt.derivative (t, o) << std::endl;
      std::cout << "Trajectory state (one call):" << std::endl
		<< ftt.state (t, orderMax) << std::endl
		<< "Trajectory state variation (splitted):" << std::endl;
      for (unsigned o = 0; o <= orderMax; ++o)
	std::cout << ftt.variationDerivWrtParam (t, o) << std::endl;
      std::cout << "Trajectory state (one call):" << std::endl
		<< ftt.variationStateWrtParam (t, orderMax) << std::endl;

      try
	{
	  std::cout << "Check frontal speed gradient." << std::endl;
	  checkGradientAndThrow (*frontalSpeed, 0, ftt.state (t, orderMax));

	  std::cout << "Check orthogonal speed gradient." << std::endl;
	  checkGradientAndThrow (*orthogonalSpeed, 0,
				 ftt.state (t, orderMax));

	  std::cout << "Check state cost gradient." << std::endl;
	  for (unsigned gradientId = 0;
	       gradientId < stateFunction.outputSize (); ++gradientId)
	    checkGradientAndThrow (stateFunction, gradientId, fttParams);
	}
      catch (BadGradient& bg)
	{
	  std::cout << bg << std::endl;
	}
      std::cout << "\n\n" << std::endl;
    }

  return 0;
}

GENERATE_TEST ()
