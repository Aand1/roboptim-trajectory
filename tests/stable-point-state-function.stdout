State cost evaluation:
[1](0.00)
State cost gradient:
[13](0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00)
Trajectory state (splitted):
[3](0.00,0.00,0.00)
[3](0.00,0.00,0.00)
Trajectory state (one call):
[6](0.00,0.00,0.00,0.00,0.00,0.00)
Trajectory state variation (splitted):
[3,13]((0.00,1.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00),(0.00,0.00,1.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00),(0.00,0.00,0.00,1.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00))
[3,13]((0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00),(0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00),(0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00))
Trajectory state (one call):
[6,13]((0.00,1.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00),(0.00,0.00,1.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00),(0.00,0.00,0.00,1.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00),(0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00),(0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00),(0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00))
Check frontal speed gradient.
Check orthogonal speed gradient.
Check state cost gradient.



State cost evaluation:
[1](5.86)
State cost gradient:
[13](3.91,-0.23,0.00,22.95,0.23,0.00,0.49,0.00,0.00,0.00,0.00,0.00,0.00)
Trajectory state (splitted):
[3](0.52,2.08,0.00)
[3](7.81,31.25,0.00)
Trajectory state (one call):
[6](0.52,2.08,0.00,7.81,31.25,0.00)
Trajectory state variation (splitted):
[3,13]((0.78,0.98,0.00,0.00,0.02,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00),(3.12,0.00,0.98,0.00,0.00,0.02,0.00,0.00,0.00,0.00,0.00,0.00,0.00),(0.00,0.00,0.00,0.98,0.00,0.00,0.02,0.00,0.00,0.00,0.00,0.00,0.00))
[3,13]((11.72,-0.31,0.00,0.00,0.31,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00),(46.88,0.00,-0.31,0.00,0.00,0.31,0.00,0.00,0.00,0.00,0.00,0.00,0.00),(0.00,0.00,0.00,-0.31,0.00,0.00,0.31,0.00,0.00,0.00,0.00,0.00,0.00))
Trajectory state (one call):
[6,13]((0.78,0.98,0.00,0.00,0.02,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00),(3.12,0.00,0.98,0.00,0.00,0.02,0.00,0.00,0.00,0.00,0.00,0.00,0.00),(0.00,0.00,0.00,0.98,0.00,0.00,0.02,0.00,0.00,0.00,0.00,0.00,0.00),(11.72,-0.31,0.00,0.00,0.31,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00),(46.88,0.00,-0.31,0.00,0.00,0.31,0.00,0.00,0.00,0.00,0.00,0.00,0.00),(0.00,0.00,0.00,-0.31,0.00,0.00,0.31,0.00,0.00,0.00,0.00,0.00,0.00))
Check frontal speed gradient.
Check orthogonal speed gradient.
Check state cost gradient.



State cost evaluation:
[1](23.44)
State cost gradient:
[13](15.62,-0.94,0.00,78.12,0.94,0.00,15.62,0.00,0.00,0.00,0.00,0.00,0.00)
Trajectory state (splitted):
[3](4.17,16.67,0.00)
[3](31.25,125.00,0.00)
Trajectory state (one call):
[6](4.17,16.67,0.00,31.25,125.00,0.00)
Trajectory state variation (splitted):
[3,13]((6.25,0.83,0.00,0.00,0.17,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00),(25.00,0.00,0.83,0.00,0.00,0.17,0.00,0.00,0.00,0.00,0.00,0.00,0.00),(0.00,0.00,0.00,0.83,0.00,0.00,0.17,0.00,0.00,0.00,0.00,0.00,0.00))
[3,13]((46.88,-1.25,0.00,0.00,1.25,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00),(187.50,0.00,-1.25,0.00,0.00,1.25,0.00,0.00,0.00,0.00,0.00,0.00,0.00),(0.00,0.00,0.00,-1.25,0.00,0.00,1.25,0.00,0.00,0.00,0.00,0.00,0.00))
Trajectory state (one call):
[6,13]((6.25,0.83,0.00,0.00,0.17,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00),(25.00,0.00,0.83,0.00,0.00,0.17,0.00,0.00,0.00,0.00,0.00,0.00,0.00),(0.00,0.00,0.00,0.83,0.00,0.00,0.17,0.00,0.00,0.00,0.00,0.00,0.00),(46.88,-1.25,0.00,0.00,1.25,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00),(187.50,0.00,-1.25,0.00,0.00,1.25,0.00,0.00,0.00,0.00,0.00,0.00,0.00),(0.00,0.00,0.00,-1.25,0.00,0.00,1.25,0.00,0.00,0.00,0.00,0.00,0.00))
Check frontal speed gradient.
Check orthogonal speed gradient.
Check state cost gradient.



State cost evaluation:
[1](46.87)
State cost gradient:
[13](31.25,-1.41,0.00,58.59,1.17,0.00,56.15,0.23,0.00,2.44,0.00,0.00,0.00)
Trajectory state (splitted):
[3](13.54,47.92,0.00)
[3](62.50,156.25,0.00)
Trajectory state (one call):
[6](13.54,47.92,0.00,62.50,156.25,0.00)
Trajectory state variation (splitted):
[3,13]((18.75,0.50,0.00,0.00,0.48,0.00,0.00,0.02,0.00,0.00,0.00,0.00,0.00),(46.88,0.00,0.50,0.00,0.00,0.48,0.00,0.00,0.02,0.00,0.00,0.00,0.00),(0.00,0.00,0.00,0.50,0.00,0.00,0.48,0.00,0.00,0.02,0.00,0.00,0.00))
[3,13]((78.12,-1.88,0.00,0.00,1.56,0.00,0.00,0.31,0.00,0.00,0.00,0.00,0.00),(-15.62,0.00,-1.88,0.00,0.00,1.56,0.00,0.00,0.31,0.00,0.00,0.00,0.00),(0.00,0.00,0.00,-1.88,0.00,0.00,1.56,0.00,0.00,0.31,0.00,0.00,0.00))
Trajectory state (one call):
[6,13]((18.75,0.50,0.00,0.00,0.48,0.00,0.00,0.02,0.00,0.00,0.00,0.00,0.00),(46.88,0.00,0.50,0.00,0.00,0.48,0.00,0.00,0.02,0.00,0.00,0.00,0.00),(0.00,0.00,0.00,0.50,0.00,0.00,0.48,0.00,0.00,0.02,0.00,0.00,0.00),(78.12,-1.88,0.00,0.00,1.56,0.00,0.00,0.31,0.00,0.00,0.00,0.00,0.00),(-15.62,0.00,-1.88,0.00,0.00,1.56,0.00,0.00,0.31,0.00,0.00,0.00,0.00),(0.00,0.00,0.00,-1.88,0.00,0.00,1.56,0.00,0.00,0.31,0.00,0.00,0.00))
Check frontal speed gradient.
Check orthogonal speed gradient.
Check state cost gradient.



State cost evaluation:
[1](70.31)
State cost gradient:
[13](46.88,-0.94,0.00,0.00,0.00,0.00,0.00,0.94,0.00,0.00,0.00,0.00,0.00)
Trajectory state (splitted):
[3](29.17,66.67,0.00)
[3](93.75,0.00,0.00)
Trajectory state (one call):
[6](29.17,66.67,0.00,93.75,0.00,0.00)
Trajectory state variation (splitted):
[3,13]((37.50,0.17,0.00,0.00,0.67,0.00,0.00,0.17,0.00,0.00,0.00,0.00,0.00),(0.00,0.00,0.17,0.00,0.00,0.67,0.00,0.00,0.17,0.00,0.00,0.00,0.00),(0.00,0.00,0.00,0.17,0.00,0.00,0.67,0.00,0.00,0.17,0.00,0.00,0.00))
[3,13]((109.37,-1.25,0.00,0.00,0.00,0.00,0.00,1.25,0.00,0.00,0.00,0.00,0.00),(-500.00,0.00,-1.25,0.00,0.00,0.00,0.00,0.00,1.25,0.00,0.00,0.00,0.00),(0.00,0.00,0.00,-1.25,0.00,0.00,0.00,0.00,0.00,1.25,0.00,0.00,0.00))
Trajectory state (one call):
[6,13]((37.50,0.17,0.00,0.00,0.67,0.00,0.00,0.17,0.00,0.00,0.00,0.00,0.00),(0.00,0.00,0.17,0.00,0.00,0.67,0.00,0.00,0.17,0.00,0.00,0.00,0.00),(0.00,0.00,0.00,0.17,0.00,0.00,0.67,0.00,0.00,0.17,0.00,0.00,0.00),(109.37,-1.25,0.00,0.00,0.00,0.00,0.00,1.25,0.00,0.00,0.00,0.00,0.00),(-500.00,0.00,-1.25,0.00,0.00,0.00,0.00,0.00,1.25,0.00,0.00,0.00,0.00),(0.00,0.00,0.00,-1.25,0.00,0.00,0.00,0.00,0.00,1.25,0.00,0.00,0.00))
Check frontal speed gradient.
Check orthogonal speed gradient.
Check state cost gradient.



State cost evaluation:
[1](82.03)
State cost gradient:
[13](54.69,-0.23,0.00,-1.95,-1.17,0.00,-44.92,1.17,0.00,-44.92,0.23,0.00,-1.95)
Trajectory state (splitted):
[3](50.00,50.00,0.00)
[3](109.38,-125.00,0.00)
Trajectory state (one call):
[6](50.00,50.00,0.00,109.38,-125.00,0.00)
Trajectory state variation (splitted):
[3,13]((54.69,0.02,0.00,0.00,0.48,0.00,0.00,0.48,0.00,0.00,0.02,0.00,0.00),(-62.50,0.00,0.02,0.00,0.00,0.48,0.00,0.00,0.48,0.00,0.00,0.02,0.00),(0.00,0.00,0.00,0.02,0.00,0.00,0.48,0.00,0.00,0.48,0.00,0.00,0.02))
[3,13]((54.69,-0.31,0.00,0.00,-1.56,0.00,0.00,1.56,0.00,0.00,0.31,0.00,0.00),(-62.50,0.00,-0.31,0.00,0.00,-1.56,0.00,0.00,1.56,0.00,0.00,0.31,0.00),(0.00,0.00,0.00,-0.31,0.00,0.00,-1.56,0.00,0.00,1.56,0.00,0.00,0.31))
Trajectory state (one call):
[6,13]((54.69,0.02,0.00,0.00,0.48,0.00,0.00,0.48,0.00,0.00,0.02,0.00,0.00),(-62.50,0.00,0.02,0.00,0.00,0.48,0.00,0.00,0.48,0.00,0.00,0.02,0.00),(0.00,0.00,0.00,0.02,0.00,0.00,0.48,0.00,0.00,0.48,0.00,0.00,0.02),(54.69,-0.31,0.00,0.00,-1.56,0.00,0.00,1.56,0.00,0.00,0.31,0.00,0.00),(-62.50,0.00,-0.31,0.00,0.00,-1.56,0.00,0.00,1.56,0.00,0.00,0.31,0.00),(0.00,0.00,0.00,-0.31,0.00,0.00,-1.56,0.00,0.00,1.56,0.00,0.00,0.31))
Check frontal speed gradient.
Check orthogonal speed gradient.
Check state cost gradient.



State cost evaluation:
[1](70.31)
State cost gradient:
[13](46.88,0.00,0.00,0.00,-0.94,0.00,0.00,0.00,0.00,0.00,0.94,0.00,0.00)
Trajectory state (splitted):
[3](70.83,33.33,0.00)
[3](93.75,0.00,0.00)
Trajectory state (one call):
[6](70.83,33.33,0.00,93.75,0.00,0.00)
Trajectory state variation (splitted):
[3,13]((56.25,0.00,0.00,0.00,0.17,0.00,0.00,0.67,0.00,0.00,0.17,0.00,0.00),(0.00,0.00,0.00,0.00,0.00,0.17,0.00,0.00,0.67,0.00,0.00,0.17,0.00),(0.00,0.00,0.00,0.00,0.00,0.00,0.17,0.00,0.00,0.67,0.00,0.00,0.17))
[3,13]((-46.87,0.00,0.00,0.00,-1.25,0.00,0.00,0.00,0.00,0.00,1.25,0.00,0.00),(750.00,0.00,0.00,0.00,0.00,-1.25,0.00,0.00,0.00,0.00,0.00,1.25,0.00),(0.00,0.00,0.00,0.00,0.00,0.00,-1.25,0.00,0.00,0.00,0.00,0.00,1.25))
Trajectory state (one call):
[6,13]((56.25,0.00,0.00,0.00,0.17,0.00,0.00,0.67,0.00,0.00,0.17,0.00,0.00),(0.00,0.00,0.00,0.00,0.00,0.17,0.00,0.00,0.67,0.00,0.00,0.17,0.00),(0.00,0.00,0.00,0.00,0.00,0.00,0.17,0.00,0.00,0.67,0.00,0.00,0.17),(-46.87,0.00,0.00,0.00,-1.25,0.00,0.00,0.00,0.00,0.00,1.25,0.00,0.00),(750.00,0.00,0.00,0.00,0.00,-1.25,0.00,0.00,0.00,0.00,0.00,1.25,0.00),(0.00,0.00,0.00,0.00,0.00,0.00,-1.25,0.00,0.00,0.00,0.00,0.00,1.25))
Check frontal speed gradient.
Check orthogonal speed gradient.
Check state cost gradient.



State cost evaluation:
[1](46.88)
State cost gradient:
[13](31.25,0.00,0.00,0.00,-0.23,0.00,2.44,-1.17,0.00,56.15,1.41,0.00,58.59)
Trajectory state (splitted):
[3](86.46,52.08,0.00)
[3](62.50,156.25,0.00)
Trajectory state (one call):
[6](86.46,52.08,0.00,62.50,156.25,0.00)
Trajectory state variation (splitted):
[3,13]((43.75,0.00,0.00,0.00,0.02,0.00,0.00,0.48,0.00,0.00,0.50,0.00,0.00),(109.37,0.00,0.00,0.00,0.00,0.02,0.00,0.00,0.48,0.00,0.00,0.50,0.00),(0.00,0.00,0.00,0.00,0.00,0.00,0.02,0.00,0.00,0.48,0.00,0.00,0.50))
[3,13]((-78.12,0.00,0.00,0.00,-0.31,0.00,0.00,-1.56,0.00,0.00,1.87,0.00,0.00),(296.88,0.00,0.00,0.00,0.00,-0.31,0.00,0.00,-1.56,0.00,0.00,1.87,0.00),(0.00,0.00,0.00,0.00,0.00,0.00,-0.31,0.00,0.00,-1.56,0.00,0.00,1.87))
Trajectory state (one call):
[6,13]((43.75,0.00,0.00,0.00,0.02,0.00,0.00,0.48,0.00,0.00,0.50,0.00,0.00),(109.37,0.00,0.00,0.00,0.00,0.02,0.00,0.00,0.48,0.00,0.00,0.50,0.00),(0.00,0.00,0.00,0.00,0.00,0.00,0.02,0.00,0.00,0.48,0.00,0.00,0.50),(-78.12,0.00,0.00,0.00,-0.31,0.00,0.00,-1.56,0.00,0.00,1.87,0.00,0.00),(296.88,0.00,0.00,0.00,0.00,-0.31,0.00,0.00,-1.56,0.00,0.00,1.87,0.00),(0.00,0.00,0.00,0.00,0.00,0.00,-0.31,0.00,0.00,-1.56,0.00,0.00,1.87))
Check frontal speed gradient.
Check orthogonal speed gradient.
Check state cost gradient.



State cost evaluation:
[1](23.44)
State cost gradient:
[13](15.62,0.00,0.00,0.00,0.00,0.00,0.00,-0.94,0.00,15.62,0.94,0.00,78.12)
Trajectory state (splitted):
[3](95.83,83.33,0.00)
[3](31.25,125.00,0.00)
Trajectory state (one call):
[6](95.83,83.33,0.00,31.25,125.00,0.00)
Trajectory state variation (splitted):
[3,13]((25.00,0.00,0.00,0.00,0.00,0.00,0.00,0.17,0.00,0.00,0.83,0.00,0.00),(100.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.17,0.00,0.00,0.83,0.00),(0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.17,0.00,0.00,0.83))
[3,13]((-109.38,0.00,0.00,0.00,0.00,0.00,0.00,-1.25,0.00,0.00,1.25,0.00,0.00),(-437.50,0.00,0.00,0.00,0.00,0.00,0.00,0.00,-1.25,0.00,0.00,1.25,0.00),(0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,-1.25,0.00,0.00,1.25))
Trajectory state (one call):
[6,13]((25.00,0.00,0.00,0.00,0.00,0.00,0.00,0.17,0.00,0.00,0.83,0.00,0.00),(100.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.17,0.00,0.00,0.83,0.00),(0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.17,0.00,0.00,0.83),(-109.38,0.00,0.00,0.00,0.00,0.00,0.00,-1.25,0.00,0.00,1.25,0.00,0.00),(-437.50,0.00,0.00,0.00,0.00,0.00,0.00,0.00,-1.25,0.00,0.00,1.25,0.00),(0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,-1.25,0.00,0.00,1.25))
Check frontal speed gradient.
Check orthogonal speed gradient.
Check state cost gradient.



State cost evaluation:
[1](5.86)
State cost gradient:
[13](3.91,0.00,0.00,0.00,0.00,0.00,0.00,-0.23,0.00,0.49,0.23,0.00,22.95)
Trajectory state (splitted):
[3](99.48,97.92,0.00)
[3](7.81,31.25,0.00)
Trajectory state (one call):
[6](99.48,97.92,0.00,7.81,31.25,0.00)
Trajectory state variation (splitted):
[3,13]((7.03,0.00,0.00,0.00,0.00,0.00,0.00,0.02,0.00,0.00,0.98,0.00,0.00),(28.12,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.02,0.00,0.00,0.98,0.00),(0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.02,0.00,0.00,0.98))
[3,13]((-66.41,0.00,0.00,0.00,0.00,0.00,0.00,-0.31,0.00,0.00,0.31,0.00,0.00),(-265.62,0.00,0.00,0.00,0.00,0.00,0.00,0.00,-0.31,0.00,0.00,0.31,0.00),(0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,-0.31,0.00,0.00,0.31))
Trajectory state (one call):
[6,13]((7.03,0.00,0.00,0.00,0.00,0.00,0.00,0.02,0.00,0.00,0.98,0.00,0.00),(28.12,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.02,0.00,0.00,0.98,0.00),(0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.02,0.00,0.00,0.98),(-66.41,0.00,0.00,0.00,0.00,0.00,0.00,-0.31,0.00,0.00,0.31,0.00,0.00),(-265.62,0.00,0.00,0.00,0.00,0.00,0.00,0.00,-0.31,0.00,0.00,0.31,0.00),(0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,-0.31,0.00,0.00,0.31))
Check frontal speed gradient.
Check orthogonal speed gradient.
Check state cost gradient.



